### PR TITLE
OFI:  move OFI_RETRY_UNTIL_DONE to common

### PR DESCRIPTION
**Pull request by hppritcha**
In running some MPI stress tests on a HPE SS11 network we had to fall back to using the OFI BTL path.  That in turn revealed some places in the BTL where we need to use a function similar to the MTL_OFI_RETRY_UNTIL_DONE macro.

So as a first step move this macro to ofi common layer and invoke the more general opal_progress function.  That's the content of this PR

Additional changes needed to the OFI BTL will applied in subsequent PRs.

The OFI MTL will require more work as the situation hit with the HPE CXI provider indicates a need to implement some kind of send backlog queueing mechanism in the MTL rather than simply spinning on the OFI CQs hoping for progress at the OFI provider level.

### CODE REVIEW

**Comment by wenduwan** [comment]
> The OFI MTL will require more work as the situation hit with the HPE CXI provider indicates a need to implement some kind of send backlog queueing mechanism in the MTL rather than simply spinning on the OFI CQs hoping for progress at the OFI provider level.

@hppritcha Could you elaborate on the problem? What is the queue needed for?

**Comment by hppritcha** [comment]
well what appears to happen is that there are some communication patterns (like each MPI process posting a large number of isends) before doing any other MPI calls that may "sink" some of these, i.e. probing or receiving, that the OFI CXI provider gets plugged up and itself can't make progress.  So our OFI MTL gets plugged up spinning on call to the ```fi_tsend``` or equivalent.  Doesn't help to call ```fi_cqread``` owing to this pluggage in the OFI CXI provider.  It appears to the application that they get "stuck" in a ```MPI_Isend```.  With the OMPI OB1 PML this isn't a problem since it has a ```add_request_to_send_pending``` method that allows for the case when sends be posted.

**Comment by wenduwan** [comment]
May I ask about the scale of the application? I want to understand if the problem is unique to CXI or common with EFA.

Is the reason understood for stuck progress?

**Comment by hppritcha** [comment]
it only takes around 400-450 MPI processes actually to observe the hang behavior.

**Comment by wenduwan** [comment]
> it only takes around 400-450 MPI processes actually to observe the hang behavior.

Hmmmm this is not particularly large in any sense. Do you have a reproducer program for me to try?

Is it possible to be a libfabric bug?

**Review by wenduwan — APPROVED** [review]
Minor formatting issue.

File: ompi/mca/mtl/ofi/mtl_ofi.h [hunk 1]
```diff
@@ -636,7 +617,7 @@ ompi_mtl_ofi_post_recv_excid_buffer(bool blocking, struct ompi_communicator_t *c
     ofi_req->completion_count = 1;
     ofi_req->comm = comm;
 
-    MTL_OFI_RETRY_UNTIL_DONE(fi_recv(ompi_mtl_ofi.ofi_ctxt[0].rx_ep,
+    OFI_RETRY_UNTIL_DONE(fi_recv(ompi_mtl_ofi.ofi_ctxt[0].rx_ep,
                                       start,
                                       length,
```
[the remark this answers is not in the document]
Comment (hppritcha) [reply ompi/mca/mtl/ofi/mtl_ofi.h:622]: oh i'll fix that

**Comment by hppritcha** [comment]
The behavior of the CXI provider (and its a little confusing because what's in the libfabric github repo https://github.com/ofiwg/libfabric/tree/main/prov/cxi/src isn't exactly what's installed on our system), indicates there's some kind of resource (send ids or something used for rendezvous) that gets exhausted until something happens on the receiver end of the message (like i said above probing for a message or posting a receive) to return one or more of these IDs.  I doubt EFA encounters this problem.  I am pretty sure this isn't a problem with the generic libfabric functionality like linked list support, etc.

There's nothing wrong per se with app's comm pattern.  Probably not the smartest way to do things.  But, it worked at scale on our omnipath (PSM2) clusters and older Cray XC/XE systems as well as on IBM power9/ib systems using spectrum MPI.

**Comment by wenduwan** [comment]
@hppritcha Thanks for shedding lights on the problem.

In that case I guess it's worthy fixing the CXI rendezvous protocol(might be harder). I feel somewhat uneasy about introducing another layer of overhead in MTL - in fact we are looking for ways to strip down indirection and let the network handle bursts.

**Comment by hppritcha** [comment]
I'd prefer to leave the OFI MTL alone as well for now - modulo this change.

**Pull request merged**